parcels/particle.py

```diff
@@ -251,6 +254,26 @@ def remove(self, indices):
         self.particles = np.delete(self.particles, indices)
         return particles
 
+    def density(self, lon=None, lat=None):
+        if lon is None:
+            lon = self.grid.U.lon
+        if lat is None:
+            lat = self.grid.U.lat
+        Density = np.zeros((lon.size, lat.size), dtype=np.float32)
+        Area = np.zeros(np.shape(Density), dtype=np.float32)
+
+        dx = (lon[1] - lon[0]) * 1852 * 60 * np.cos(lat*math.pi/180)
+        dy = (lat[1] - lat[0]) * 1852 * 60
+        for y in range(len(lat)):
+            Area[0:len(lon), y] = dy * dx[y]
+
+        # For each particle, find closest vertex in x and y
+        for p in self.particles:
+            Density[np.argmin(np.abs(p.lon - lon)), np.argmin(np.abs(p.lat - lat))] += 1
+        # Scale by cell area
```

Comment:
It turns out `particle.xi` and `yi` are not updated each time-step, so we need to do this binning somewhere every time we want to calculate densities anyway.